### PR TITLE
Unit-dependent recall costs

### PR DESCRIPTION
**Pull request by Slayer95**
Gold calculated after reviewing the builds I used in my TLU playthroughs (though I don't have any in Easy).

### CODE REVIEW

**Comment by Slayer95** [comment]
61719558ab makes me doubtful. I really need more player data (replays) for Siege of Kharos.

File: scenarios/16_Defense_of_Saffaros.cfg [hunk 1]
```diff
@@ -605,6 +625,7 @@ Note that you can toggle these effects at any time during the scenario by using
                 bonus=no
                 carryover_percentage=40
             [/gold_carryover]
+            note=_"Do not capture Dedylos' villages"
```
Comment (Slayer95) [inline scenarios/16_Defense_of_Saffaros.cfg]: Yet Dedylos seems very interested in capturing our villages...

File: scenarios/16_Defense_of_Saffaros.cfg [hunk 2]
```diff
@@ -79,11 +81,13 @@ Note that you can toggle these effects at any time during the scenario by using
 
         recruit=EoMa_White_Warrior,EoMa_Protector,EoMa_Brown_Warrior,EoMa_Shielder,EoMa_Cavalry_Archer,EoMa_Sun_Follower,EoMa_Adept_of_Light
 
+        ai_special=guardian
```
Comment (Slayer95) [inline scenarios/16_Defense_of_Saffaros.cfg]: Fails to enforce Dedylos' staying in his keep :confused:

**Comment by Slayer95** [comment]
Update on this:

So, according to my observation, the campaign may be divided in three blocks regarding gold accumulation.
- Block 1. S01-S05. Everything up to Mehir getting Sud-Affar's gold. Gold in these scenarios doesn't need updating.
- Block 2. S06-S15. Because of Southern Nations' carryover, starting gold for S06 is predictable. After striking it rich in Ka-gatta, gold is quickly drained in each of the following scenarios, making it hard to keep any carryover.
- Block 3. S16-S23. Previously, S19 drained all the gold available. However, after #28, there is now a small amount of gold that carries over from S19 to S22.

## Block 1
Both in **S03 Near the Jungle** and **S04 Southern Nations**, the option of recalling Lv1 troops at 16-17 gold is increasingly attractive. This makes leveling up your original army from S01 Highest Tiers a much more interesting option. This results in making their traits (which are random) much more important down the road, since they will often decide whether they become Summoners or Banishers (my rule of thumb is that everyone should have 6 MP, which means Quick=Summoner, but I guess others have their own preferences).

In **S04 Southern Nations**, starting gold is very limited, so now being able to recall cheap Lv1 units make it potentially easier (on the other hand, recalling 1 Summoner or Dispeller at 23g compensates 1 Lv1 unit at 17g.) Getting Rhaxiz to survive isn't exactly trivial, so being easier is fine.

## Block 2
In **S06 Fire Canyon**, there are likely now some veterans in our army. A small boost of 10g helps support it, and/or it may be used to summon a Rhami (with S04's Heavy Summoner), or an Earth/Fire Elemental should we have a Grand Summoner.

While **S08 Ka-gatta** is a defying scenario in a first gameplay, it's no longer threatening upon replaying. Don't get fooled by inflation: while there is a boost of 10g here, too, it actually makes the scenario a bit harder.

This scenario used to have the parameters
```
village_support=4
village_income=2
```

However, ``village_income`` doesn't exist. It should be ``village_gold=2``. And, since this scenario doesn't need to be easier, ``village_support`` gets lowered in compensation. Overall, this means that if the player uses a _small squad_ to assault this scenario, they will get more carryover gold as a reward. If they actually do that, then the scenario will be harder for them, which is good.

After S08, player should be rich. So, gold in **S09 Goddess of All Rhamis** has not been touched. This exacerbates the draining nature _Block 2_, however.

**S10 Chaos** depends a lot in the composition of the player army and their willingness to take risks. Adding just 30g means that the player will not be able to recall a heavy assault force and be fearless against whatever gets out of the Red Abyss.

**S12 Siege of Mag-Magar** is a bit peculiar in that there is a huge amount of units involved, but so little space. This forces the player to prioritize recalling high-level units rather than cheap low-level units. Given the scale of the battle, it turns out that 70g in S12 is a modest amount. Note that after the update that transfers control of allied forces to the player, this scenario was made easier. So much more gold isn't really needed.

**S15 Abyssal Rebellion** also gets 70g to support an army capable of resisting the assault of high-level reds.

## Block 3
**S16 Defense of Saffaros** is left outside of this PR. But it's to be noted that this scenario takes place in mostly open field, so recruiting low-level units en masse is viable here.

**S17 Siege of Kharos** is harder, specially so because of the new turn limit. Income is balanced accounting for the player taking over ally villages.

**S19 Skydock** gets further gold to increase the player's chances of getting carryover.

**S22 North Pole** and **S23 Ruins** match expectations as originally discussed in #26 .

**Comment by inferno8** [comment]
I've been playtesting this PR for a few days and I am very impressed. The new unit-dependent recall costs system changes everything. This is such a game changer that it makes the infamous jinn ban completely redundant - and I am going to remove it. With this in mind, along with my playthrough I've been implementing new jinn advice for earlier scenarios. Some of them can completely alter entire scenario progression. This is especially noticeable in S10 "Chaos" where the player can now complete the mission with the portal staying closed all the time.

I am currently on S12 (normal mode) and I am playing with a minor jinn cost modification to make these special units even more expensive to recall (formula+4g). Still, even with the increased costs, I find the overall progression a bit easier, but maybe this is just me. I need to play more to verify my observations.
Once I complete my current playthrough, I'll review this PR again.

**Comment by Slayer95** [comment]
> This is especially noticeable in S10 "Chaos" where the player can now complete the mission with the portal staying closed all the time.

If I am visualizing this correctly, that should be by sending a bunch of Mystical Jinn through the sides in an assassination mission? Baiting Aerius with Wonderful Jinn might also work, though he can be easily prevented from moving. However, if the plan is assassination, Air Gods should do the job better, since they are as fast as Mystical Jinn but get Skirmisher. In any case, control of this map's flanks with aerial forces is very important, since it even provides protection against the portal's pull. So, maybe Aerius should get a few Cosmic Eyes in order to better compete?

> I find the overall progression a bit easier, but maybe this is just me. I need to play more to verify my observations.

If djinn are available, I have no doubt that gold needs to be restricted further: ``4g`` of djinn tax is probably not enough, since, as you mention, they can entirely change how scenarios work, mainly because of their mobility. Getting djinn in S12 -well known for being hard to maneuver in- is truly a godsend.

**Comment by Slayer95** [comment]
I have pushed my proposal for S16 here, lest it misses your testing, @inferno8 . It basically makes the ally AI fight cohesively, but attack the wrong targets, causing its army to perish. This in agreement with later's Mehir observation that Kharos soldiers fare poorly against Tharis. Tharis is also boosted so that they can recruit many more units.

**Comment by Slayer95** [comment]
Regarding djinn in S12, it may be claimed that, since they belong to Sud-Affar's army, they will be sent to Mag-Magar as reinforcement together with the rest of the Trinity's army. But of course those reinforcements will not arrive.

**Comment by inferno8** [comment]
> If I am visualizing this correctly, that should be by sending a bunch of Mystical Jinn through the sides in an assassination mission?

One of the most amazing facts about Wonderful and Mystical Jinn in TLU universe is that they can foresee the future (in a limited way but still). In this particular scenario if the player has such a Jinni under his command, the Jinni simply tells Mehir that Rashti'rhami is going to be tricked by Aerius into opening the portal. Knowing this, Mehir asks her to not do this. There's more to that, but I don't want to go into much detail right now.

> `4g` of djinn tax is probably not enough, since, as you mention, they can entirely change how scenarios work

At the moment I am leaning towards `+6g` or even `*1.5` but I need to play more. Btw, I am thinking about keeping the overcost even after Mehir becomes Summons Master to avoid Jinn recall spam in late-game.
Below is the current wip note about the Jinn overcost. The note is displayed after getting a Neutral Summoner:
`Note: Your unit has become a Neutral Summoner. He can now summon friendly Jinn to aid your cause. Jinn, while incredibly useful, are rare and very expensive to maintain. This means recalling them in the next scenarios will be costly.`
The "expensive maintenance" could be explained further by the fact that Jinn require lots of books to satisfy their thirst for knowledge, etc.

> Regarding djinn in S12, it may be claimed that, since they belong to Sud-Affar's army they will be sent to Mag-Magar as reinforcement together with the rest of the Trinity's army. But of course those reinforcements will not arrive.

I don't want to exclude them in this scenario. Actually this mission is still challenging even with Jinn available to recall. With a proper recall cost modification I think it is possible to achieve good balance there.

**Comment by Slayer95** [comment]
> One of the most amazing facts about Wonderful and Mystical Jinn in TLU universe is that they can foresee the future (in a limited way but still). In this particular scenario if the player has such a Jinni under his command, the Jinni simply tells Mehir that Rashti'rhami is going to be tricked by Aerius into opening the portal. Knowing this, Mehir asks her to not do this. There's more to that, but I don't want to go into much detail right now.

Interesting! Looking forward to it.

> At the moment I am leaning towards +6g or even *1.5 but I need to play more. Btw, I am thinking about keeping the overcost even after Mehir becomes Summons Master to avoid Jinn recall spam in late-game.

That sounds good. It would also help at improving the transition from S22 to S23, where you keep all your Lv4 summoners for free. So the djinn overcost would encourage the player to get more summoners and less djinn.

> The "expensive maintenance" could be explained further by the fact that Jinn require lots of books to satisfy their thirst for knowledge, etc.

Right. Even from the prologue they are seen in need of money to hire the ferryman, lol.

> I don't want to exclude them in this scenario. Actually this mission is still challenging even with Jinn available to recall. With a proper recall cost modification I think it is possible to achieve good balance there.

Fair enough. It's not like the player has no airforce alternatives, too.

**Comment by Slayer95** [comment]
> At the moment I am leaning towards +6g or even *1.5 but I need to play more. Btw, I am thinking about keeping the overcost even after Mehir becomes Summons Master to avoid Jinn recall spam in late-game.

Something that may also work is a modest increase in recall cost accompanied with +1g of extra upkeep cost. That would fit closer with the narrative of "more expensive to maintain", and it'd be less visually scary. Paying 5g per turn for a Mystical Jinn would also be arguably less painful than paying 2g per turn for a common Jinn, while being effectively a similar gold penalty.

**Comment by inferno8** [comment]
> Something that may also work is a modest increase in recall cost accompanied with +1g of extra upkeep cost. That would fit closer with the narrative of "more expensive to maintain", and it'd be less visually scary.

To be honest, I am not a big fan of upkeep modifications. Once implemented they would probably add another layer of complexity to the game, which is already complex in my opinion. But I am not saying no to implementing your idea in the future, if things like `formula+7g` fail. 

The `+4g` for recalling Jinn is definitely too low. S15 Abyssal Rebellion was really easy to win because of that. I am about to playtest S16 Defense of Saffaros with `+7g` now.

**Comment by inferno8** [comment]
After playtesting more than 2/3 of the campaign, I think the PR is ready for merging. The new system/mechanic will require a separate recall cost formula for Jinn, which once officially implemented will have to be thoroughly tested and adjusted. I will commit my proposal very soon.

Btw, the unit-dependent recall cost idea is one of the most significant changes in the history of the add-on and makes early Jinn a possibility. Well done @Slayer95 and thank you.

Lets continue the discussion in #26 and/or in new PRs, once this is closed.

**Pull request merged**